src/devices/bus/isa/hpblp.cpp

```diff
@@ -0,0 +1,555 @@
+	m_tms9914->int_write_cb().set(FUNC(isa8_hpblp_device::gpib_irq));
+
+	IEEE488(config, m_ieee488, 0);
+	m_ieee488->eoi_callback().set(m_tms9914, FUNC(tms9914_device::eoi_w));
+	m_ieee488->dav_callback().set(m_tms9914, FUNC(tms9914_device::dav_w));
+	m_ieee488->nrfd_callback().set(m_tms9914, FUNC(tms9914_device::nrfd_w));
+	m_ieee488->ndac_callback().set(m_tms9914, FUNC(tms9914_device::ndac_w));
+	m_ieee488->ifc_callback().set(m_tms9914, FUNC(tms9914_device::ifc_w));
+	m_ieee488->srq_callback().set(m_tms9914, FUNC(tms9914_device::srq_w));
+	m_ieee488->atn_callback().set(m_tms9914, FUNC(tms9914_device::atn_w));
+	m_ieee488->ren_callback().set(m_tms9914, FUNC(tms9914_device::ren_w));
+
+	ieee488_slot_device &slot0(IEEE488_SLOT(config, "ieee0", 0));
+	hp_ieee488_devices(slot0);
+	slot0.set_default_option("hp9122c");
+
+	bus::hp_dio::dio16_device &dio16(DIO16(config, "diobus", 0));
+	dio16.set_program_space(m_maincpu, AS_PROGRAM);
+	m_maincpu->reset_cb().set(dio16, FUNC(bus::hp_dio::dio16_device::reset_in));
+
+	dio16.irq1_out_cb().set(FUNC(isa8_hpblp_device::dio_irq1_w));
+	dio16.irq2_out_cb().set(FUNC(isa8_hpblp_device::dio_irq2_w));
+	dio16.irq3_out_cb().set(FUNC(isa8_hpblp_device::dio_irq3_w));
+	dio16.irq4_out_cb().set(FUNC(isa8_hpblp_device::dio_irq4_w));
+	dio16.irq5_out_cb().set(FUNC(isa8_hpblp_device::dio_irq5_w));
+	dio16.irq6_out_cb().set(FUNC(isa8_hpblp_device::dio_irq6_w));
+	dio16.irq7_out_cb().set(FUNC(isa8_hpblp_device::dio_irq7_w));
+
+	DIO32_SLOT(config, "sl0", 0, "diobus", dio16_cards, nullptr, false);
+	DIO32_SLOT(config, "sl1", 0, "diobus", dio16_cards, nullptr, false);
+
+}
+
+offs_t isa8_hpblp_device::get_bus_address(offs_t offset, uint16_t mem_mask) const
+{
+	if (mem_mask == 0xff)
+		offset++;
+	return offset;
+}
+
+bool isa8_hpblp_device::forward_to_host(offs_t offset) const
+{
+	if (offset >= 0x200000 && offset <= 0x2fffff)
+		return true;
+	if (offset >= 0x420000 && offset <= 0x42ffff)
+		return true;
+	if (offset >= 0x510000 && offset <= 0x51ffff)
+		return true;
+	if (offset >= 0x520000 && offset <= 0x52ffff)
+		return true;
+	if (offset >= 0x538000 && offset <= 0x53ffff)
+		return true;
+	if (offset >= 0x5f2000 && offset <= 0x5f3fff)
+		return true;
+	if (offset >= 0x690000 && offset <= 0x69ffff)
+		return true;
+	if (offset >= 0x6f0000 && offset <= 0x6fffff)
+		return true;
+	if (offset >= 0x730000 && offset <= 0x73ffff)
+		return true;
+	if (offset >= 0x7a0000 && offset <= 0x7affff)
+		return true;
+	return false;
+}
+
+uint8_t isa8_hpblp_device::status_val(offs_t offset) const
+{
+	if (offset >= 0x200000 && offset <= 0x2fffff)
+		return 0x30;
+	if (offset >= 0x512000 && offset <= 0x517fff)
+		return 0x70;
+	if (offset >= 0x51ff00 && offset <= 0x51ffff)
+		return 0x50;
+	if (offset >= 0x538000 && offset <= 0x53ffff)
+		return 0xb0;
+	return 0xd0;
+}
+
+uint16_t isa8_hpblp_device::bus_r(offs_t offset, uint16_t mem_mask)
+{
+	offset <<= 1;
+
+	if (machine().side_effects_disabled())
+		return 0xffff;
```

Comment:
Would it be better to return `m_bus_data` if it isn’t an address that triggers a bus error so you at least see something in the debugger?